From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Wed, 3 Jan 2024 21:00:41 +0100
Subject: [PATCH] Client perspective - Base ids on perspective - Block -
 Provide

Provides functionality to base `Block` ids (the ids they are identified by in packets sent to clients) on the target client's perspective.
For clients that can understand the server-side block ids directly, the id is taken from the regular registry (`BuiltInRegistries.BLOCK`), and for those who do not, the ids are taken from the vanilla-only registry (`Block.VANILLA_ONLY_BLOCK_REGISTRY`).

Does not add any usages of this functionality.

License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)
Fiddle - https://fiddlemc.org

diff --git a/src/main/java/org/fiddlemc/fiddle/packet/ClientPerspective.java b/src/main/java/org/fiddlemc/fiddle/packet/ClientPerspective.java
index b97f5798056d3083cb0ba2e11c55544cff9dde81..eb26df34a458d48319fe7500322525661e06a0fb 100644
--- a/src/main/java/org/fiddlemc/fiddle/packet/ClientPerspective.java
+++ b/src/main/java/org/fiddlemc/fiddle/packet/ClientPerspective.java
@@ -59,6 +59,43 @@ public class ClientPerspective {
         this.understandsServerSideTranslations = understandsServerSideTranslations;
     }
 
+    // Fiddle start - client perspective - base ids on perspective - provide - block
+    /**
+     * @return The {@link net.minecraft.core.IdMap} that
+     * {@linkplain net.minecraft.core.IdMap#getId provides} the ids for
+     * {@linkplain net.minecraft.world.level.block.Block blocks}
+     * that a client with this perspective interprets correctly.
+     */
+    public net.minecraft.core.IdMap<net.minecraft.world.level.block.Block> getPacketBlockIdMap() {
+        return understandsServerSideItemsAndBlocks ? net.minecraft.core.registries.BuiltInRegistries.BLOCK : net.minecraft.world.level.block.Block.VANILLA_ONLY_BLOCK_REGISTRY;
+    }
+
+    /**
+     * @return The id provided by this perspective's {@link #getPacketBlockIdMap()},
+     * for the given {@link net.minecraft.world.level.block.Block}.
+     * A null value will be interpreted as
+     * {@link net.minecraft.world.level.block.Blocks#AIR}.
+     * <p>
+     *     This is based on {@link #getPacketBlockIdMap()}, but uses the cached id for optimization instead.
+     * </p>
+     */
+    public int getPacketBlockId(@org.jetbrains.annotations.Nullable net.minecraft.world.level.block.Block block) {
+        var blockToQuery = block == null ? net.minecraft.world.level.block.Blocks.AIR : block;
+        return understandsServerSideItemsAndBlocks ? blockToQuery.indexInBlockRegistry : blockToQuery.indexInVanillaOnlyBlockRegistry;
+    }
+
+    /**
+     * @return The single {@link #getPacketBlockId} value for the given
+     * {@link net.minecraft.world.level.block.Block}, if it is the same
+     * for all possible perspectives, or -1 otherwise.
+     */
+    public static int getSinglePacketBlockId(@org.jetbrains.annotations.Nullable net.minecraft.world.level.block.Block block) {
+        var blockToQuery = block == null ? net.minecraft.world.level.block.Blocks.AIR : block;
+        int potentialSingleIndex = blockToQuery.indexInBlockRegistry;
+        return potentialSingleIndex == blockToQuery.indexInVanillaOnlyBlockRegistry ? potentialSingleIndex : -1;
+    }
+    // Fiddle end - client perspective - base ids on perspective - provide - block
+
     @Override
     public int hashCode() {
         return this.category.ordinal();
